policystore/store.go

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2018 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package policystore
+
+import (
+	"sync"
+
+	"github.com/projectcalico/app-policy/proto"
+	"github.com/prometheus/common/log"
+)
+
+type PolicyStore struct {
+	// The RWMutex protects the entire contents of the PolicyStore. No one should read from or write to the PolicyStore
+	// without acquiring the corresponding lock.
+	// Helper methods Write() and Read() encapsulate the correct locking logic.
+	RWMutex sync.RWMutex
+
+	PolicyByID  map[proto.PolicyID]*proto.Policy
+	ProfileByID map[proto.ProfileID]*proto.Profile
+	IPSetByID   map[string]IPSet
+	Endpoint    *proto.WorkloadEndpoint
+}
+
+func NewPolicyStore() *PolicyStore {
```

Comment:
just FYI: returning a pointer here would cause allocation in the heap which can be avoided if we use `func NewPolicyStore(p *PolicyStore)` instead since it will be stored in the stack instead of the heap. But this is a premature performance tuning unless we have performance issues.

Reply:
Good to know.